relayer/src/supervisor.rs

```diff
@@ -379,3 +435,42 @@ pub fn collect_events(batch: EventBatch) -> CollectedEvents {
 
     collected
 }
+
+// TODO: Memoize this result
+fn get_counterparty_chain(
+    src_chain: &dyn ChainHandle,
+    src_channel_id: &ChannelId,
+    src_port_id: &PortId,
+) -> Result<ChainId, BoxError> {
+    info!(
+        chain_id = %src_chain.id(),
+        src_channel_id = %src_channel_id,
+        src_port_id = %src_port_id,
+        "getting counterparty chain"
+    );
+
+    use ibc::ics02_client::client_state::ClientState;
+
+    let src_channel = src_chain.query_channel(src_port_id, src_channel_id, Height::zero())?;
+
+    // TODO: Check channel state?
+
+    let src_connection_id = src_channel
+        .connection_hops()
+        .first()
+        .ok_or_else(|| format!("no connection hops for channel '{}'", src_channel_id))?;
+
+    let src_connection = src_chain.query_connection(&src_connection_id, Height::zero())?;
+
+    // TODO: Check connection state?
```

Comment:
@adizere @ancazamfir Same here, do we need to check the connection state to infer the corresponding object or does it not matter?

Reply:
Just looked the query for non-existing connection returns a connection in `Uninitialized` state but with `07-tendermint-0` as client id (??). So I think for both channel and connection we should check that the state is not `Uninitialized` before continuing.